reco_utils/recommender/sar/sar_singlenode.py

```diff
@@ -50,533 +28,331 @@ class SARSingleNode:
-
-        return (
-            results[[self.col_user, self.col_item, self.col_rating]]
-            .rename(columns={self.col_rating: PREDICTION_COL})
-            .astype(
-                {
-                    self.col_user: _user_item_return_type(),
-                    self.col_item: _user_item_return_type(),
-                    PREDICTION_COL: self.scores.dtype,
-                }
-            )
+        # ensure datatypes are correct
+        df = df.astype(
+            dtype={
+                self.col_user: str,
+                self.col_item: str,
+                self.col_prediction: self.scores.dtype,
+            }
         )
 
+        # drop seen items
+        return df.replace(-np.inf, np.nan).dropna()
+
     def predict(self, test):
         """Output SAR scores for only the users-items pairs which are in the test set
-
         Args:
-            test (pd.DataFrame): DataFrame that contains ground-truth of user-item ratings.
-
+            test (pd.DataFrame): DataFrame that contains users and items to test
         Return:
-            pd.DataFrame: DataFrame contains the prediction results.
+            pd.DataFrame: DataFrame contains the prediction results
         """
-        # pick users from test set and
-        test_users = test[self.col_user].unique()
-        try:
-            training_ids = np.array([self.user_map_dict[user] for user in test_users])
-            assert training_ids is not None
-        except KeyError():
-            msg = "SAR cannot score test set users which are not in the training set"
-            log.error(msg)
-            raise ValueError(msg)
-
-        # shorthand
-        scores = self.scores
-
-        # Convert to dense, the following operations are easier.
-        log.info("Converting to dense array ...")
-        scores_dense = scores.toarray()
-
-        # take the intersection between train test items and items we actually need
-        test_col_hashed_users = test[self.col_user].map(self.user_map_dict)
-        test_col_hashed_items = test[self.col_item].map(self.item_map_dict)
-
-        test_index = pd.concat(
-            [test_col_hashed_users, test_col_hashed_items], axis=1
-        ).values
-        aset = set([tuple(x) for x in self.index])
-        bset = set([tuple(x) for x in test_index])
-
-        common_index = np.array([x for x in aset & bset])
-
-        # Mask out items in the train set.  This only makes sense for some
-        # problems (where a user wouldn't interact with an item more than once).
-        if self.remove_seen and len(aset & bset) > 0:
-            log.info("Removing seen items...")
-            scores_dense[common_index[:, 0], common_index[:, 1]] = 0
-
-        final_scores = scores_dense[test_index[:, 0], test_index[:, 1]]
-
-        results = pd.DataFrame.from_dict(
+
+        # get user / item indices from test set
+        user_ids = test[self.col_user].map(self.user2index).values
+        assert not any(
```

Comment:
instead of having an assert, can you throw an exception?

Reply:
why are exceptions preferred? (I'm fine changing it, just curious about the difference)

Reply:
it is a good practice that we are following in the rest of the code, that way we can control how to catch them in case it is necessary. 

Normally what we do is, throw exceptions for code and use asserts for tests

Reply:
ok, np